Mark applying block templates not persistent

draganescu opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

The same issue appeared twice:

- [once](https://github.com/WordPress/gutenberg/pull/44596) when building a fallback for the navigation block
- [once](https://github.com/WordPress/gutenberg/pull/45776) when building a page list block that expands in list view

In both cases, we’re setting the template for a block asynchronously and that creates a persistent change in the post which seems to be because of [how templated blocks work](https://github.com/WordPress/gutenberg/blob/fecc1d2aa66c2fb2c6379fb47de5d7cd71f01f8c/packages/block-editor/src/components/inner-blocks/use-inner-block-template-sync.js#L84-L96)

This PR updates that to make sure applying templates does not create an undo level.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Applying templates to block inner contents should not create undo levels because by being automattic it is decent to assume the content is already saved and it will not depend on the initial state of the template.


## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->


Using a simple `__unstableMarkNextChangeAsNotPersistent` call before applying the template.

## Testing Instructions

Easiest way to test is to checkout #45776 which implements this change.
Another way is to check out this PR and try to set block templates to custom blocks, core blocks, and also set the template async.

## Screenshots or screencast <!-- if applicable -->

N/A

draganescu committed:
mark applying templates as a not persistent change

youknowriad commented:
Well the e2e tests here give some sense of confidence I think. Do you think there's a test we should add here?

talldan commented:
> Well the e2e tests here give some sense of confidence I think. Do you think there's a test we should add here?

Could be good to check for regressions of the issue this fixes. Possibly a template that's initially empty and then populated after a `setTimeout`.

Not sure how easy it is to construct a test like that, not sure what the existing template coverage is like.

draganescu commented:
@talldan wouldn't any test involving the Page List block after #45776 basically test this?

Mamaduka commented:
It might be a good time to check existing template test coverage. For example, do we have tests that track if the template synchronization works correctly?

P.S. I'll add e2e tests for #45632 to avoid similar regressions in the future.

talldan commented:
> @talldan wouldn't any test involving the Page List block after https://github.com/WordPress/gutenberg/pull/45776 basically test this?

Only if those tests assert that the post has unsaved changes. I guess it'd be ok to add a more specific page list test for this.

draganescu commented:
Ok so I managed to get to this repro code:

<details>
<summary> Expand for code </summary>

```JS
const el = wp.element.createElement;
const useState = wp.element.useState;
const InnerBlocks = wp.blockEditor.InnerBlocks;
const TEMPLATE_TWO_PARAGRAPHS = [
	[
		'core/paragraph',
		{
			fontSize: 'large',
			content: 'One',
		},
	],
	[
		'core/paragraph',
		{
			fontSize: 'large',
			content: 'Two',
		},
	],
];

window.wp.blocks.registerBlockType( 'test/test-inner-blocks-async-template', {
	title: 'Test Inner Blocks no locking',
	icon: 'cart',
	category: 'text',

	edit() {

		const [ template, setTemplate ] = useState([]);

		setInterval( () => {
			setTemplate( TEMPLATE_TWO_PARAGRAPHS );
		}, 1000 );
	
		return el( InnerBlocks, {
			template: template,
		} );
	},

	save() {
		return el( InnerBlocks.Content );
	},
} );
	
	
const testBlock = window.wp.blocks.createBlock ( 'test/test-inner-blocks-async-template', {} )
window.wp.data.dispatch( 'core/block-editor' ).insertBlock( testBlock );

```

</details>

Pasting this and running it in the console:

- on `trunk` produces two undo levels
- on this branch `fix/async-block-templates` produces one undo level

How do best implement this as a test?

If the console test is good, and in the interest of speed (I'll be away for a bit) if anyone can contribute the above code as a test to this branch I'd be ❤️ 

- add the block to `packages/e2e-tests/plugins/inner-blocks-templates/index.js`
- add a test adding this block and counting undo levels somewhere (not sure where this should be? `test/e2e/specs/editor/various/block-locking.spec.js` or a new suite?). 

Otherwise I'll pick it up back when I return.

talldan commented:
The regression test would be great, but also IMO, the main thing to manually test before merging this is that it doesn't break other situations where templates are used.

It's pretty easy to assert that this solves the known situation, but it'd be good to get a handle on whether there are situations that we don't know about by looking at other usages of templates in the codebase.

The main one is - could there be a situation in which the content should be marked as dirty, but this change breaks that. Looking at the code it's unlikely, but some reassurance would be good.

youknowriad commented:
> could there be a situation in which the content should be marked as dirty, but this change breaks that. Looking at the code it's unlikely, but some reassurance would be good.

Can't think of any either but it's hard to be certain know before actually shipping.

scruffian committed:
Add a test

scruffian commented:
@draganescu I tried adding a test, but it didn't work - my tests pass in trunk as well as in this branch. I added the code as a commit to show how far I've got in case you (or anyone else) is able to pick it up...

talldan committed:
Update e2e test

- Move block code to test plugin file
- Move from block locking test file to a new inner blocks template file
- Update test logic to test post is not dirty on a fresh load
- Rename post type templates file to distinguish from inner blocks templates file

talldan commented on test/e2e/specs/editor/various/inner-blocks-templates.spec.js:
```diff
@@ -0,0 +1,53 @@
+test.describe( 'Inner blocks templates', () => {
+	test.beforeAll( async ( { requestUtils } ) => {
+		await requestUtils.activatePlugin(
+			'gutenberg-test-inner-blocks-templates'
```
This test plugin seems to have been unused. Not sure how that happened. Probably a good chance to cull much of its code, but I won't do that in this PR.

talldan committed:
Fix tests

- Avoid saving inner blocks of the test block, as otherwise template sync won't run
- Wait for template resolution before saving, as the test seems flakey otherwise

talldan commented:
> @draganescu I tried adding a test, but it didn't work - my tests pass in trunk as well as in this branch. I added the code as a commit to show how far I've got in case you (or anyone else) is able to pick it up...

I've pushed a commit that updates the test, and it should work as expected now. I think the main thing missing was saving the post, reloading and then checking the undo count and publish buttons aren't enabled. The other issue was that the test block was saving its inner blocks, which meant template sync wasn't being run.

Mamaduka commented on test/e2e/specs/editor/various/inner-blocks-templates.spec.js:
```diff
@@ -0,0 +1,56 @@
+			'role=document[name="Block: Test Inner Blocks Async Template"i] >> text=OneTwo'
+		);
+
+		// The block template content appears asynchronously, so wait for it.
+		await blockWithTemplateContent.waitFor();
```
Nit: I think this would be the same as the assertion below. I personally prefer more explicit assertions.

```js
await expect( blockWithTemplateContent ).toBeVisible();
```

scruffian committed:
make more explicit assetions

scruffian reviewed (approved):
✅  tests pass locally
✅  tests fail locally when I remove the fix
✅  code looks good
